server/block/redstone.go

```diff
@@ -0,0 +1,394 @@
+			r.RedstoneUpdate(pos, w)
+		}
+	}
+}
+
+// updateDirectionalRedstone updates redstone components through the given face. This implementation is based off of
+// RedstoneCircuit and Java 1.19.
+func updateDirectionalRedstone(pos cube.Pos, w *world.World, face cube.Face) {
+	updateAroundRedstone(pos, w)
+	updateAroundRedstone(pos.Side(face), w, face.Opposite())
+}
+
+// receivedRedstonePower returns true if the given position is receiving power from any faces that aren't ignored.
+func receivedRedstonePower(pos cube.Pos, w *world.World, ignoredFaces ...cube.Face) bool {
+	for _, face := range cube.Faces() {
+		if slices.Contains(ignoredFaces, face) {
+			continue
+		}
+		if w.RedstonePower(pos.Side(face), face, true) > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// identifyNeighbours identifies the neighbouring positions of a given node, determines their types, and links them into
+// the graph. After that, based on what nodes in the graph have been visited, the neighbours are reordered left-to-right
+// relative to the direction of information flow.
+func (n *wireNetwork) identifyNeighbours(w *world.World, node *wireNode) {
+	neighbours := computeRedstoneNeighbours(node.pos)
+	neighboursVisited := make([]bool, 0, 24)
+	neighbourNodes := make([]*wireNode, 0, 24)
+	for _, neighbourPos := range neighbours[:24] {
+		neighbour, ok := n.nodeCache[neighbourPos]
+		if !ok {
+			neighbour = &wireNode{
+				pos:   neighbourPos,
+				block: w.Block(neighbourPos),
+			}
+			n.nodeCache[neighbourPos] = neighbour
+			n.nodes = append(n.nodes, neighbour)
+		}
+
+		neighbourNodes = append(neighbourNodes, neighbour)
+		neighboursVisited = append(neighboursVisited, neighbour.visited)
+	}
+
+	fromWest := neighboursVisited[0] || neighboursVisited[7] || neighboursVisited[8]
+	fromEast := neighboursVisited[1] || neighboursVisited[12] || neighboursVisited[13]
+	fromNorth := neighboursVisited[4] || neighboursVisited[17] || neighboursVisited[20]
+	fromSouth := neighboursVisited[5] || neighboursVisited[18] || neighboursVisited[21]
+
+	var cX, cZ int32
+	if fromWest {
+		cX++
+	}
+	if fromEast {
+		cX--
+	}
+	if fromNorth {
+		cZ++
+	}
+	if fromSouth {
+		cZ--
+	}
+
+	var heading uint32
+	if cX == 0 && cZ == 0 {
+		heading = computeRedstoneHeading(node.xBias, node.zBias)
+		for _, neighbourNode := range neighbourNodes {
+			neighbourNode.xBias = node.xBias
+			neighbourNode.zBias = node.zBias
+		}
+	} else {
+		if cX != 0 && cZ != 0 {
+			if node.xBias != 0 {
+				cZ = 0
+			}
+			if node.zBias != 0 {
+				cX = 0
+			}
+		}
+		heading = computeRedstoneHeading(cX, cZ)
+		for _, neighbourNode := range neighbourNodes {
+			neighbourNode.xBias = cX
+			neighbourNode.zBias = cZ
+		}
+	}
+
+	n.orientNeighbours(&neighbourNodes, node, heading)
+}
+
+// reordering contains lookup tables that completely remap neighbour positions into a left-to-right ordering, based on
+// the cardinal direction that is determined to be forward.
+var reordering = [][]uint32{
```

Comment:
Worth considering this is in the block package so a more redstone-specific name might need to be considered